rfcs/text/0015_nodejs_clustering.md

```diff
@@ -0,0 +1,509 @@
+- Start Date: 2021-03-09
+- RFC PR: https://github.com/elastic/kibana/pull/94057
+- Kibana Issue: https://github.com/elastic/kibana/issues/68626
+- POC issue: https://github.com/elastic/kibana/pull/93380
+
+# Summary
+
+Leveraging NodeJS clustering API to have multi-processes Kibana instances.
+
+# Motivation
+
+Kibana currently uses a single Node process/thread to serve HTTP traffic.
+This means Kibana cannot take advantage of multi-core hardware making it expensive to scale out Kibana.
+
+Clustering mode would allow spawning multiple Kibana processes ('workers') from a single Kibana instance.
+
+This should be an optional way to run Kibana since users running Kibana inside docker containers might
+choose to rather use their container orchestration to run a container (with a single kibana process)per host CPU.
+
+# Architecture summary
+
+In 'classic' mode, the Kibana server is started in the main NodeJS process.
+
+![image](../images/15_clustering/no-cluster-mode.png)
+
+In clustering mode, the main NodeJS process would only start the coordinator, which would then fork workers using 
+the cluster API. NodeJS underlying socket implementation allows multiple processes to listen to the same ports, 
+performing http traffic balancing between the workers for us.
+
+![image](../images/15_clustering/cluster-mode.png)
+
+The coordinator sole responsibility is to orchestrate the workers. It would not be a 'super' worker handling both 
+the job of a worker while being in charge of managing the other workers.
+
+# Performance testing
+
+## Local testing
+
+These tests were performed against a local development machine, with a 8-core cpu (2.4 GHz 8-Core Intel Core i9 - 32 GB 2400 MHz DDR4),
+using the default configuration of the `kibana-load-testing` tool.
+
+### Raw results
+
+#### Non-clustered mode
+
+![image](../images/15_clustering/perf-no-clustering.png)
+
+#### Clustered mode, 2 workers
+
+![image](../images/15_clustering/perf-clustering-2-worker.png)
+
+#### Clustered mode, 4 workers
+
+![image](../images/15_clustering/perf-4-workers.png)
+
+### Analysis 
```

Comment:
Maybe you could explain a bit more how our benchmark test works? My understanding is that they perform a single flow and measure the response times for completing that flow. This is a good smoke test to see what performance a single user would observe on a mostly idle system. However, does our benchmarking run enough parallel requests to start maxing out a single and even more so, multiple core's?

Unless our benchmark is causing the node process to consume so much CPU that we start seeing delays in the event-loop (and some requests start hitting a timeout) we're unlikely to see the true performance gain clustering could achieve.

Although faster response time is a good sign that Kibana is able to handle more, we probably want to measure the request per second throughput.

Reply:
> Although faster response time is a good sign that Kibana is able to handle more, we probably want to measure the request per second throughput.

Yea, this is not how gatling works tbh. The best we can do would be increasing the number of performed requests per batch.

Reply:
I don't know anything about our setup but it seems like gattling should be able to give us requests succeeded vs requests failed (OK vs KO) https://gatling.io/docs/current/general/reports/

If we can get this metric we should push concurrent users up until we start seeing at least some failures (like maybe 1-5%).

We would then have to double the concurrent users when trying with two workers and compare the number of successful requests against the one worker scenario.

Reply:
FWIW, when doing performance testing for Fleet, I had to do what @rudolf suggested and manually increase the concurrent users to find the breaking point.